Adds a new salvager weapon: The Makeshift Laser Rifle

unfunnyperson opened the pull request:
[FEATURE] [GAME OBJECTS] [INPUT] [WIKI]

## About the PR 

Adds a new salvager weapon!
![image](https://github.com/goonstation/goonstation/assets/77701409/463bc7b4-9419-41dc-9a52-d01fe84141ed)

The weapon is a 2 handed laser weapon.
It fits on backs as well.
It uses large energy cells and uses 1250 energy per shot to create a 20 damage laser.

different cells used
![image](https://github.com/goonstation/goonstation/assets/77701409/d2d7a98a-ac7f-4c76-966e-7d5b466d5527)
![image](https://github.com/goonstation/goonstation/assets/77701409/00c1b1ad-cd90-43af-85e4-03b0cd775ffd)

The weapon can be used in-hand to adjust the firerate. This is between 0.2 and 1 seconds
![image](https://github.com/goonstation/goonstation/assets/77701409/473c0506-6ad7-4996-8aa4-1aa96160c2dc)

If the gun is fired too much within a short period of time or the gun is emped the gun will burst into flame and the gun's cell will be drained!
![image](https://github.com/goonstation/goonstation/assets/77701409/31d7d6e2-6049-42e7-99cd-d69b4fbaa551)
After cooling down the gun needs to be repaired
![image](https://github.com/goonstation/goonstation/assets/77701409/402db7c2-0b65-4131-9028-c0085cec1869)
Steps:
1. Wirecutters to remove burnt wires
2. Add 10 wires to replace burnt wires
3. remove and replace power cell/light tube

The meltdown combined with the adjustable firerate allows a player to choose a slower firerate for a more reliable gun or a faster firerate for higher dps.

Using a rigged light tube will cause the gun to explode on use.

Using a wrench on the rifle will remove the battery
Using wirecutters will remove the light tube

A new battery/light tube can be added to the gun by clicking on it with the respective item

## Why's this needed? <!-- Describe why you think this should be added to the game. -->

More variety in salvager weapons provides a more interesting gameplay experience. This weapon is designed around being less powerful than the survival rifle but has more ammo sources than the rifle as compensation. 

## Changelog

```changelog
(u)UnfunnyPerson
(+)Added a new salvager weapon, the makeshift laser rifle.
```

unfunnyperson commented:
Still need to add help messages and a sprite for the potato cell

pgmzeta commented:
it would be cool if it used the matsci parts like the captain's gun does

CheffieGithub commented:
On testing I think you could up the damage, on the fastest fire rate it was able to fire 7 times for 84 damage. On the slowest 12 times for 144 damage. Maybe change the temperature scaling a bit as well?

Considering the RP-3 micro phaser is incredibly easy to get and can do more damage in about the same span of time with 10 damage and 20 shots with dual wield capabilities, pocket storage and easier cell swapping.

CheffieGithub commented on code/obj/item/gun/energy.dm:
```diff
@@ -1858,3 +1858,276 @@ TYPEINFO(/obj/item/gun/energy/wasp)
+			return
+		..()
+
+	examine()
+		. = ..()
+		if (!our_cell)
+			. += "[src] is lacking a power source!"
```
I would use `get_desc()` here instead and add cases for missing / broken tube.

CheffieGithub replied:
there should also be a space before `[src]` here

unfunnyperson commented:
Changed damage from 12 >18
Added a new construction step to make construction a little more difficult
The gun overheating is now a lot worse and need repairs

imawxome commented:
i really like this, its great to have some usefull makeshift guns

pali6 commented:
I personally am not convinced we want something like a makeshift laser gun in the game.

Azrun commented on code/obj/item/gun/energy.dm:
```diff
@@ -1839,3 +1839,428 @@ TYPEINFO(/obj/item/gun/energy/wasp)
+	///What step of repair are we on if we have broken? 0 = functional
+	var/heat_repair = 0
+	///What projectile we switch to when update_proj() is called
+	var/lens_proj = /datum/projectile/laser/makeshift
```
Why is this necessary given the existing members on this type?

unfunnyperson replied:
fixed

unfunnyperson commented:
reopening this because I want to make this a weapon for salvagers instead of a makeshift crew weapon

unfunnyperson commented:
don't have enough time to fix this right now, ill try to get it done on Friday or Thursday

Azrun commented on code/obj/item/gun/energy.dm:
```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
+		our_cell.set_loc(src)
+		our_cell.AddComponent(/datum/component/power_cell, our_cell.maxcharge, our_cell.charge, our_cell.genrate, 0, FALSE)
+		SEND_SIGNAL(src, COMSIG_CELL_SWAP, our_cell)
+		update_icon()
```
```suggestion
		UpdateIcon()
```

Azrun commented on code/obj/item/gun/energy.dm:
```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
+	Exited(Obj, newloc)
+		var/obj/item/cell/C = Obj
+		if (istype(C) && !QDELETED(C))
+			C.update_icon()
```
```suggestion
			C.UpdateIcon()
```

Azrun commented on code/obj/item/gun/energy.dm:
```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
+				playsound(src, 'sound/items/Wirecutter.ogg', 50, TRUE)
+				user.put_in_hand_or_drop(our_light)
+				our_light = null
+				update_icon()
```
```suggestion
				UpdateIcon()
```

Azrun commented on code/obj/item/gun/energy.dm:
```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
+				boutput(user,SPAN_NOTICE("You place [W] inside of the barrel and redo the wiring."))
+				playsound(src, 'sound/effects/pop.ogg', 50, TRUE)
+				attach_light(W, user)
+				update_icon()
```
```suggestion
				UpdateIcon()
```

Azrun commented on code/obj/item/gun/energy.dm:
```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
+				var/obj/item/cell/artifact/C = our_cell
+				var/datum/artifact/powercell/AS = C.artifact
+				var/datum/artifact_origin/AO = AS.artitype
+				overlay_image = SafeGetOverlayImage("gun_cell", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-[AO.name]")
+			else
+				overlay_image = SafeGetOverlayImage("gun_cell", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-[our_cell.icon_state]")
+			src.UpdateOverlays(overlay_image, "gun_cell")
+		else
+			src.UpdateOverlays(null, "gun_cell")
+
+		if (our_light)
+			var/image/overlay_image = SafeGetOverlayImage("gun_light", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-light")
+			src.UpdateOverlays(overlay_image, "gun_light")
+		else
+			src.UpdateOverlays(null, "gun_light")
+
+		if (heat > FIRE_THRESHOLD)
+			var/image/overlay_image = SafeGetOverlayImage("gun_smoke", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-burn")
+			src.UpdateOverlays(overlay_image, "gun_smoke")
+		else if (heat > 70)
+			var/image/overlay_image = SafeGetOverlayImage("gun_smoke", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-smoke")
```
Instances of `'icons/obj/items/guns/energy64x32.dmi'` can be replaced by src.icon

Azrun commented on code/obj/item/gun/energy.dm:
```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
+				attach_cell(W, user)
+				return
+			else if (issnippingtool(W) && our_light)
+				boutput(user,SPAN_NOTICE("You remove [our_light] from the barrel."))
```
I'd prefer this refer to cutting away the wiring attaching the light.  Otherwise I don't understand how you remove it with a snipping implement.

Azrun commented:
Should light tube adjust projectile color?

unfunnyperson commented:
> > Should light tube adjust projectile color?
> 
> That would be cool but I can't figure out how to do it
Nevermind I did find a way to do this 


> > Should light tube adjust projectile color?
> 
> That would be cool but I can't figure out how to do it
figured out a way
